### ISSUE TITLE
Error: Could not find a suitable provider for docker_compose

### ISSUE THREAD

**Issue by marrajo-afk** [puppetlabs/puppetlabs-docker#742]
## Describe the Bug
When use classes docker and docker::compose in an own class (lamp) , the first puppet agent execution failed with:
**Error: Could not find a suitable provider for docker_compose**

It only works when the **docker-ce** package is installed manually in the node before the first puppet agent is run in it.

`lamp/init.pp:`
  include docker
  include docker::compose

  Class['docker::compose'] ->
  Class['docker'] ->
  File['docker-compose'] ->
  Docker_compose['lamp']

  file { "docker-compose":
          name    => "/var/pr1/docker-compose.yml",
          replace => true,
          content => template("lamp/docker-compose.erb"),
          notify  => Docker_Compose['lamp']
       }
  docker_compose { 'lamp':
      compose_files => ['/var/pr1//docker-compose.yml'],
      ensure              => present,
  }

If i change the order of execution, the result is the same, i.e:
  Class['docker'] ->
  Class['docker::compose'] ->
  ....
  ....

`docker/manifest/params.pp:`
    $compose_install_path   = '/usr/bin'
    $compose_version        = '1.29.1'

After the failed execution, both docker-compose executable and docker-ce packages are correctly installed

## Expected Behavior
Correct execution of docker_compose Resource Type

## Environment
 Puppet master: 
 Foreman 2.2
 Debian 10
 module puppetlabs-docker (v4.0.0)
 Puppet v7.5.0

Agent:
Debian10
Puppet v7.5.0

Thanks.

**Comment by benh57** [comment]
Also seeing this, seems the provider doesnt use the full path to docker-compose, which is in /usr/local/bin. (when managed by this module), but /usr/local/bin is not in the default PATH for puppet on my servers.

**Comment by hugoboos** [comment]
Quick fix by linking `docker-compose` to `/usr/bin`.

```puppet
class { '::docker::compose':
  ensure  => present,
  version => '1.29.2',
}

file { '/usr/bin/docker-compose':
  ensure  => link,
  target  => '/usr/local/bin/docker-compose',
  require => Class['::docker::compose'],
}

docker_compose { 'test':
  ...
  require => File['/usr/bin/docker-compose'],
}
```

**Comment by marrajo-afk** [comment]
I force the docker-compose path in `docker/manifest/params.pp` as:
$compose_install_path = '/usr/bin'
$compose_version = '1.29.1'

After (failed) applying catalog:
$ which docker-compose
/usr/bin/docker-compose

So no luck

**Comment by markri** [comment]
I use puppet to provision a temporary AWS EC2 instance (which will be destroyed after tests have finished). So in my case it is essential that it works on the first run. However I get this same error message

- I'm sure it's a clean machine (using official CentOS 7 AMI image), so docker-ce hasn't been previously installed
- If a run a subsequent run then it works fine. So it almost looks like the docker module isn't aware of the fresh installed docker-compose yet when bringing up a docker-compose stack in the same run
- I tried suggestions above (creating a symlink at /usr/bin/docker-compose and also at /bin/docker-compose), but without satisifying result. Maybe it's not related to the PATH environment variable?

Maybe the provider could be instructed to use an absolute path instead of relying on the PATH environment variable, although at this point I'm not confident it might fix my issue. Would be happy to test though..

It's quite cumbersome to debug as I need to re-create an EC2 instance for each attempt to be able to reproduce the issue.

**Comment by mendhak** [comment]
@markri or anyone else, did you ever find a solution? I'm getting the same thing on Ubuntu 20.04, also on EC2.  Same behaviors - clean install (destroy and recreate EC2), I have to run puppet apply twice, symlink doesn't work. 

The only crude workaround I've found is to install Docker _and_ Docker-Compose in the user-data init script. Before puppet ever runs.   But this defeats the purpose of using this module. 


```
echo "Install Docker"
curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -
add-apt-repository \
   "deb [arch=amd64] https://download.docker.com/linux/ubuntu \
   $(lsb_release -cs) \
   stable"
apt-get update
apt-get install -y docker-ce
usermod -aG docker ubuntu

echo "Install latest Docker Compose"
echo "TMPDIR=/var/tmp" >> /etc/environment
COMPOSE_VERSION=`git ls-remote https://github.com/docker/compose | grep refs/tags | grep -oE "[0-9]+\.[0-9][0-9]+\.[0-9]+$" | sort --version-sort | tail -n 1`
curl -L https://github.com/docker/compose/releases/download/${COMPOSE_VERSION}/docker-compose-`uname -s`-`uname -m` > /usr/local/bin/docker-compose
chmod +x /usr/local/bin/docker-compose
```

**Comment by canihavethisone** [comment]
I use the sed script below to fix this issue, noting that it uses fixed Linux paths (not windows).  I attempted to use an exec within the has_command thus, but it returns nil:
` has_command(:dockercompose, `which docker-compose`) do`

```
#!/bin/bash

# The following sed actions will patch an issue with the puppetlabs-docker module that requires 2 puppet runs

# Change path to your modules dir
target_path='/etc/puppetlabs/code/environments/production/modules'

## --- DO NOT EDIT BELOW THIS LINE --- ##

target_file_1='docker/lib/puppet/provider/docker_compose/ruby.rb'
target_file_2='docker/lib/puppet/provider/docker_network/ruby.rb'
target_file_3='docker/lib/puppet/provider/docker_stack/ruby.rb'
target_file_4='docker/lib/puppet/provider/docker_volume/ruby.rb'

find_1="has_command(:docker, command(:dockercmd)) do"
replace_1="has_command(:docker, '/usr/bin/docker') do"

find_2="has_command(:docker_compose, command(:dockercompose)) do"
replace_2="has_command(:dockercompose, '/usr/local/bin/docker-compose') do"


sed -i "s;$find_1;$replace_1;g" \
    $target_path/{$target_file_1,$target_file_2,$target_file_3,$target_file_4}

sed -i "s;$find_2;$replace_2;g" \
    $target_path/$target_file_1
```

I still find that a second puppet run is required when `docker::selinux_enabled: true` as it performs `/etc/systemd/system/docker.service.d/service-overrides.conf]/seltype: seltype changed 'systemd_unit_file_t' to 'container_unit_file_t'` and also on any docker service files in `/etc/systemd/system/`

**Comment by canihavethisone** [comment]
I have created a pull request at https://github.com/puppetlabs/puppetlabs-docker/pull/833.  Can other users please test and verify these changes?  Thanks.

**Comment by chelnak** [comment]
Closed by #833
